src/python/pants/base/specs.py

```diff
@@ -192,6 +192,24 @@ def make_glob_patterns(self, address_mapper):
     ]
 
 
+_specificity = {
+  SingleAddress: 0,
+  SiblingAddresses: 1,
+  AscendantAddresses: 2,
+  DescendantAddresses: 3
+}
+
+
+def more_specific(spec1: Spec, spec2: Spec) -> Spec:
+  """Returns which of the two specs is more specific.
+
+  This is useful when a target matches multiple specs, and we want to associate it with
+  the "most specific" one, which will make the most intuitive sense to the user.
+  """
+  # Note that if either of spec1 or spec2 is None, the other will be returned.
+  return spec1 if _specificity.get(type(spec1), 99) < _specificity.get(type(spec2), 99) else spec2
```

Comment:
```suggestion
  return spec1 if _specificity[type(spec1)] < _specificity[type(spec2)] else spec2
```

I'd rather we fix the code if it becomes wrong, than it silently having high priority... I don't _think_ these are extensible? But if they are, we should make sure the extension hook forces setting this.

Reply:
We need these to work for `None`, but I guess I could add `NoneType` to the dict.